<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

# !!! ДАННЫЕ ДЛЯ ВХОДА УДАЛЕНЫ !!!

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса. в этом юните мной принято решение считать количество уникальных значений по столбцу с идентификатором, но возможен и вариант с подсчетом по всем столбцам.
query_3_1 = f'''select COUNT(id)
                from vacancies
            '''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''select COUNT(id)
                from employers
            '''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''select COUNT(id)
                from areas
            '''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''select COUNT(id)
                from industries
            '''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

### Выводы по предварительному анализу данных: ###
Из данного анализа видно, что в базе данных хранится информация о вакансиях от 23501 работодателя, зарегистрированных в 1362 регионах/городах, и работающих в 294 индустриях.
Это значит, что многие работодатели из базы являются прямыми конкурентами, а некоторые из них еще и работают на одной и той же территории. 
Помимо этого, в базе хораниться 49197 вакансий, следовательно в некоторые организации одновременно требуется более одного сотрудника.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса. Запросы в юните, на мой взгляд, довольно просты и не нуждаются в отдельных комментариях.
# при подсчете и именовании таблиц, руководствовался теми же принципами, что и в предыдущем юните.
query_4_1 = f'''select a.name area, COUNT(v.id) cnt
                from vacancies v
                left join areas a on v.area_id = a.id
                group by a.name
                order by 2 DESC 
            '''

In [13]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0             Москва  5333
1    Санкт-Петербург  2851
2              Минск  2112
3        Новосибирск  2006
4             Алматы  1892
..               ...   ...
764       Тарко-Сале     1
765    Новоаннинский     1
766            Бирск     1
767           Сасово     1
768             Анна     1

[769 rows x 2 columns]

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''select COUNT(id)
                from vacancies
                where (salary_from is not null) or (salary_to is not null)
            '''

In [15]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''select round(AVG(salary_from)) avg_from, round(AVG(salary_to)) avg_to
                from vacancies
            '''

In [17]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


avg_from    avg_to
0   71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''select schedule, employment, count(id) cnt
                from vacancies
                group by 1,2
                order by 3 desc
            '''

In [19]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = f'''select experience as "Требуемый опыт работы", count(id) cnt
                from vacancies
                group by 1
                order by 2
            '''

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


Требуемый опыт работы    cnt
0           Более 6 лет   1337
1             Нет опыта   7197
2         От 3 до 6 лет  14511
3    От 1 года до 3 лет  26152

***

### Выводы по детальному анализу вакансий: ###
Из типовой части юнита мы выяснили где больше выкансий, узнали пожелания работодателей.
 
Лидерами по вакансиям ожидаемо оказались крупные города под предводительством Москвы, но было приятно увидеть название своего родного городка, пусть и где-то в хвосте списка с всего одной вакансией. 

Далеко не все работодатели горят желанием сообщить соискателю сколько ему будут платить, а это признак, который тяжело спрогнозировать и автозаполнить константами или медианами, так как мы не можем залезть в бухгалтерию каждого работодателя и выяснить его платежеспособность (может быть и можем, но это очень трудозатратно, и не факт, что такую информацию можно найти в свободном доступе). А исключать эти признаки нельзя, так как они являются одними из основных для наших клиентов. Так что работаем с тем что есть.

Большая часть работодателей ищет людей готовых сидеть в офисе с 8:00 до 18:00 5 дней в неделю (если я правильно понимаю, что полный день - это непосредственное присутствие на территории работодателя), что б эти люди уже что-то умели, но еще не выгорели и не потеряли интерес к работе, и судя из данных по зарплате платить им хотят в среднем от 71 000 до 110 000 рублей.



# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса. Данный запрос можно было бы написать другим образом, создав две отдельнык таблицы с первой
# и пятой позицией и склейкой их через UNION, но запрос бы получился неоправданно длинным с дублированием многих строк. 
# Помимо этого запрос выглядит довольно синтетическим и мне трудно придумать с какой целью, 
# нам понадобился работодатель с конкретной позиции списка.
query_5_1 = f'''select e.name, count(v.id) cnt
                from vacancies v
                left join employers e on v.employer_id=e.id
                group by 1
                order by 2 desc
                limit 5
            '''

In [23]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса. С этим заданием у меня возникли проблемы. Мой изначальный код был таким:
#query_5_2 = f'''select a.name, count(distinct e.id) employers, count(distinct v.id) vacancies
#                from areas a 
#                left join employers e on a.id=e.area
#                left join vacancies v on v.area_id=e.area
#                group by 1
#                order by 3, 2 desc
#            '''
# И теоритически, он полностью выполнял задание, но при запуске запроса в следующем блоке, результата не давал, 
# а просто бесконечно выполнялся (максимум я прождал пол часа). В чем причина бесконечной обработки запроса так и не понял. Проверял и в VSCode 
# и на сайте https://www.metabase.com/ причем если убрать из SELECT distinct для обеих агрегатных функций, то результат выдает, но считает, 
# естесственно, не правильно. Приходится добавлять условие where v.id is null и код начинает работать, но информативность третьего столбца сразу 
# пропадает - там одни 0. И как быть если мы захотим посмотреть количество работодателей в регионе, где есть открытые вакансии не понятно.
query_5_2 = f'''select a.name, count(distinct e.id) employers, count(distinct v.id) vacancies
                from areas a 
                left join employers e on a.id=e.area
                left join vacancies v on v.area_id=e.area
                where v.id is null
                group by 1
                order by 3, 2 desc
            '''

In [25]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\3341481329.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  employers  vacancies
0                   Россия        410          0
1                Казахстан        207          0
2       Московская область         75          0
3       Краснодарский край         19          0
4                 Беларусь         18          0
..                     ...        ...        ...
889                Жабинка          0          0
890  Железногорск-Илимский          0          0
891                Жетысай          0          0
892               Жигалово          0          0
893               Жирновск          0          0

[894 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
query_5_3 = f'''select e.name employer, count(distinct v.area_id) cnt
                from vacancies v
                left join employers e on v.employer_id=e.id
                group by 1
                order by 2 desc
            '''

In [27]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


employer  cnt
0                      Яндекс  181
1                  Ростелеком  152
2                  Спецремонт  116
3      Поляков Денис Иванович   88
4                   ООО ЕФИН    71
...                       ...  ...
14761                  UniSol    1
14762            UNISTORY LLC    1
14763                   UNIT6    1
14764     United Distribution    1
14765              UNITEL LLC    1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
query_5_4 = f'''select count(distinct e.id)
                from employers e
                left join employers_industries s on e.id=s.employer_id
                where s.industry_id is null
            '''

In [29]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса
query_5_5 = f'''select e.name, count(s.industry_id)
                from employers e
                left join employers_industries s on e.id=s.employer_id
                group by 1
                having count(s.industry_id)=4
                order by 1
                offset 2
                limit 1
            '''

In [31]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6 = f'''select count(distinct s.employer_id)
                from employers_industries s
                left join industries i on i.id=s.industry_id
                where i.name = 'Разработка программного обеспечения' 
            '''

In [33]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников. Определять индекс таблицы на сайте пришлось методом подбора, 
# как определить его из кода не понял, но к счастью, долго подбирать не пришлось.
import urllib.parse as req

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров')
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)

cities=pd.read_html(url)[1]
cities=tuple(cities['Город'])
display(cities)

('Саратов', 'Тюмень')

In [35]:
# текст запроса. Для выполнения задания пришлось дублировать объединение таблиц датасета, возможно есть более элегантное решение, но это тоже работает.
query_5_7 = f'''select a.name, count(v.id) cnt
                from vacancies v
                left join areas a on v.area_id=a.id
                left join employers e on v.employer_id=e.id
                where e.name = 'Яндекс' and a.name in {cities}
                group by a.name
                
                union
                select 'Total', count(v.id)
                from vacancies v
                left join areas a on v.area_id=a.id
                left join employers e on v.employer_id=e.id
                where e.name = 'Яндекс' and a.name in {cities}
                order by 2
                
            '''

In [36]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  cnt
0   Тюмень   20
1  Саратов   23
2    Total   43

***

### Выводы по по анализу работодателей: ###
Задания данного юнита позволили нам узнать какие компании лидируют по количеству вакансий в нашем датасете. К нашему счастью, у лидеров списка скорее всего найдется работа для наших клиентов, так как они либо работают непосредственно в ИТ, либо имеют большие объемы данных для обработки, либо и то, и другое!

Так же мы узнали, что 3553 из 23501 работодателей нашего датасета занимаются разработкой ПО, что тоже открывает большие перспективы трудоустройства для клиентов агенства.

Помимо этого, выяснилось, что лидером по охвату территорий является Яндекс и его вакансии представленны в большем объеме во всех городах-миллионниках.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
# текст запроса. В данном запросе я применил превидение к нижнему регистру для исключения пропусков вакансний, в описании которых использован верхний регистр.
query_6_1 = f'''select count(id)
                from vacancies
                where lower(name) like '%data%' or lower(name) like '%данн%'
            '''

In [38]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
# текст запроса. А в данном случае использовал метод ilike для решения проблемы с регистрами. Сам запрос не сложный, но громоздкий из-за большого количества условий.
query_6_2 = f'''select count(distinct id)
                from vacancies
                where ((name ilike '%data scientist%'
                or name ilike '%data science%' 
                or name ilike '%исследователь данных%'
                or name ilike '%machine learning%'
                or name ilike '%машинн%обучен%')
                or (name not like '%HTML%' and name like '%ML%'))
                and (name ilike '%junior%' 
                or experience ilike '%нет опыта%' 
                or employment ilike '%стажировка%')
            '''

In [40]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
# текст запроса. В этом и следующем запросах меняются только дополнительные условия.
query_6_3 = f'''select count(distinct id)
                from vacancies
                where ((name ilike '%data scientist%'
                or name ilike '%data science%' 
                or name ilike '%исследователь данных%'
                or name ilike '%machine learning%'
                or name ilike '%машинн%обучен%')
                or (name not like '%HTML%' and name like '%ML%'))
                and (key_skills ilike '%SQL%'
                or key_skills ilike '%postgres%')
            '''

In [42]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
# текст запроса
query_6_4 = f'''select count(distinct id)
                from vacancies
                where ((name ilike '%data scientist%'
                or name ilike '%data science%' 
                or name ilike '%исследователь данных%'
                or name ilike '%machine learning%'
                or name ilike '%машинн%обучен%')
                or (name not like '%HTML%' and name like '%ML%'))
                and (key_skills ilike '%Python%')
            '''

In [44]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
# текст запроса. Использована агрегатная функция среднего значения к вычислению изменения длины строки при замене определенного символа пустой строкой.
query_6_5 = f'''select round(avg(length(key_skills)-length(replace(key_skills, CHR(9), ''))+1), 2)
                from vacancies
                where ((name ilike '%data scientist%'
                or name ilike '%data science%' 
                or name ilike '%исследователь данных%'
                or name ilike '%machine learning%'
                or name ilike '%машинн%обучен%')
                or (name not like '%HTML%' and name like '%ML%'))
                and key_skills is not null
                order by 1
            '''

In [46]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [47]:
# текст запроса. Та же функция усреднения, но теперь в сочетании с coalesce, использованной для замены пропущенных значений.
query_6_6 = f'''select experience, round(avg((coalesce(salary_from, salary_to)+coalesce(salary_to, salary_from))/2), 0) as avg_salary
                from vacancies
                where ((name ilike '%data scientist%'
                or name ilike '%data science%' 
                or name ilike '%исследователь данных%'
                or name ilike '%machine learning%'
                or name ilike '%машинн%обучен%')
                or (name not like '%HTML%' and name like '%ML%'))
                and (salary_from is not null or salary_to is not null)
                group by experience   
            '''

In [48]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

### Выводы по предметному анализу: ###
Предметный анализ позволил существенно сократить пул рассматриваемых вакансий (до 1771 вакансии) и конкретизировать требования к соискателям. Сложнее всего найти работу для соискателя без опыта работы, всего 2,8 % вакансий для соискателей в сфере Data Science содержат информацию о готовности к приему на работу специалистов без опыта.

Мы узнали, что для 20% вакансий одним из ключевых требований является владение языком Python, навыки в SQL - для 11 %, а в среднем работодатель заинтересован в соискателях с обширными знаниями (в среднем владение 6-7 ключевыми навыками).

Из данных о предлагаемых зарплатах можно сделать вывод, что в данной сфере доход работников очень быстро прогрессирует, а опытные специалисты получают в 3-4 раза больше новичков.

***

# Общий вывод по проекту

Дополнительный запрос 1. Так как агенство, на которое я работаю согласно заданию, представляет интересы соискателей, а не работодателей, интересно будет узнать в каком регионе у соискателя больший выбор работодателей. Я хотел посмотреть не только сколько работодателей выставили вакансии в данный момент, но и сколько потенциальных работодателей для ИТ специалистов имеется в регионе. По-этому я разделил запрос на два с целью дальнейшего объединения. 

In [49]:
# текст запроса. Первый запрос делает выборку регионов, в которых в данный момент открыты вакансии для наших
# клиентов и выводит их списком по уменьшению количества потенциальных работодателей


query_7_1 = f'''select a.name as "название региона", COUNT(distinct v.employer_id) as "с открытыми вакансиями"
                from areas a
                left join vacancies v on a.id=v.area_id
                where ((v.name ilike '%data scientist%'
                or v.name ilike '%data science%' 
                or v.name ilike '%исследователь данных%'
                or v.name ilike '%machine learning%'
                or v.name ilike '%машинн%обучен%')
                or (v.name not like '%HTML%' and v.name like '%ML%'))
                group by 1
                order by 2 desc
            '''

In [50]:
# по результатам запроса мы формируем первую часть нашего DataFrame и создаем кортеж с названиями регионов, попавших в выборку.
df1 = pd.read_sql_query(query_7_1, connection)
regions = tuple(df1['название региона'])

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\433627486.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df1 = pd.read_sql_query(query_7_1, connection)


In [51]:
# Второй запрос формирует выборку с количеством потенциальных работодателей для ИТ специалистов в регионах из первой выборки.
# Для этого запроса мне понадобилось соединить четыре таблицы из датасета, хоть в результате выведена информация только из двух, 
# остальные понадобились для связки и введения условий отбора. Условия сформированы субъективно на моем ощущении в каких отрослях
# наиболее востребованы ИТ специалисты. Для реальной ситуации этот список может быть дополнен и изучен подробнее.
query_7_2 = f'''select a.name as "название региона", COUNT(distinct e.id) as "работодатели в ИТ сфере"
                from areas a
                left join employers e on a.id=e.area
                left join employers_industries s on e.id=s.employer_id
                left join industries i on s.industry_id=i.id
                where (i.name like '%ИТ%'
                or i.name ilike '%интернет%' 
                or i.name like 'Разработка программного обеспечения'
                or i.name like 'Мобильная связь'
                or i.name like 'Банк')
                and a.name in {regions}
                group by 1
                order by 2 desc
            '''

In [52]:
# формируем вторую часть таблицы
df2 = pd.read_sql_query(query_7_2, connection)

# объединяем в общую таблицу
merged_df=df1.merge(
    df2,
    how='left',
    on='название региона'
)
# заменяем пропуски на 0 и выводим результат
merged_df=merged_df.fillna(0)
merged_df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\744256437.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df2 = pd.read_sql_query(query_7_2, connection)


название региона  с открытыми вакансиями  работодатели в ИТ сфере
0                      Москва                     130                   1739.0
1             Санкт-Петербург                      47                    678.0
2                 Новосибирск                      17                    166.0
3                      Казань                      13                    133.0
4                      Алматы                      11                    262.0
5             Нижний Новгород                       9                     95.0
6                       Минск                       9                    515.0
7                Екатеринбург                       6                    134.0
8                       Томск                       6                     57.0
9                   Краснодар                       6                     81.0
10                      Пермь                       5                     62.0
11             Ростов-на-Дону                       5                     91.0
12                 Нур-Султан                       4                    123.0
13                     Самара                       4                     73.0
14                       Омск                       3                     40.0
15                    Ташкент                       3                    147.0
16                     Сербия                       3                      0.0
17                       Кипр                       3                     15.0
18                    Армения                       3                      7.0
19                     Рязань                       3                     19.0
20                    Тбилиси                       3                      8.0
21                  Иннополис                       2                      5.0
22                      Чехия                       2                      4.0
23  Киров (Кировская область)                       2                     29.0
24                Владивосток                       2                     21.0
25                     Турция                       2                      1.0
26                    Воронеж                       2                     60.0
27                        Уфа                       2                     35.0
28                      Пенза                       2                     25.0
29           Набережные Челны                       1                      7.0
30                       Баку                       1                      2.0
31                    Барнаул                       1                     31.0
32                     Батуми                       1                      1.0
33                   Белгород                       1                     18.0
34                     Бишкек                       1                     44.0
35                     Брянск                       1                     16.0
36                  Волгоград                       1                     34.0
37                 Зеленоград                       1                      5.0
38                     Ижевск                       1                     48.0
39                    Израиль                       1                      0.0
40                    Королев                       1                      2.0
41                 Красноярск                       1                     38.0
42                      Литва                       1                      3.0
43                    Абхазия                       1                      0.0
44                       Орел                       1                     10.0
45               Петрозаводск                       1                      9.0
46                     Польша                       1                      1.0
47                    Саратов                       1                     26.0
48                      Тверь                       1                     17.0
49                     Тюмень                       1          

Из результатов запроса видно, что наибольшее количество вакансий и работодателей сконцентрированно в Москве. Это было ожидаемо, но не стоит упускать из виду факт, что многие компании предпочитают регистрировать головной офис в Москве, что может немного искажать результат. Помимо Москвы выделяются крупные города и столицы государств. Эти города наиболее подхоят для переезда, особенно если соискатель не хочет зависить от одного работодателя и предполагает возможность смены места работы в будущем.

Дополнительный запрос 2. В 4 юните мы определили чего хотят работодатели, но какой график работы наиболее выгоден нашим клиентам? Сформируем запрос с графиками работы и типами занятости, а также добавим к нему усредненные показатели минимальной и максимальной зарплаты и посчитаем количество вакансий по каждой строке.

In [53]:
query_4_6 = f'''select schedule, employment, round(AVG(salary_from)) avg_from, round(AVG(salary_to)) avg_to, count(id) cnt
                from vacancies
                where ((name ilike '%data scientist%'
                or name ilike '%data science%' 
                or name ilike '%исследователь данных%'
                or name ilike '%machine learning%'
                or name ilike '%машинн%обучен%')
                or (name not like '%HTML%' and name like '%ML%'))
                group by 1,2
                order by 3 desc, 4 desc
            '''

In [54]:
# результат запроса
df = pd.read_sql_query(query_4_6, connection)
df

C:\Users\andrv\AppData\Local\Temp\ipykernel_24572\934868634.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_6, connection)


schedule           employment  avg_from    avg_to  cnt
0     Гибкий график  Частичная занятость       NaN       NaN    4
1       Полный день           Стажировка       NaN       NaN    2
2  Удаленная работа     Проектная работа       NaN   90000.0    2
3     Гибкий график           Стажировка       NaN   37000.0    2
4  Удаленная работа     Полная занятость  176850.0  242138.0  110
5       Полный день     Полная занятость  140763.0  228936.0  319
6     Гибкий график     Полная занятость  140000.0  200000.0   37
7  Удаленная работа  Частичная занятость  119000.0       NaN    3
8       Полный день  Частичная занятость   45000.0  100000.0    1

Из дополнительного запроса мы можем понять, что для удовлетворения интересов наших клиентов, нам стоит обратить внимание на удаленную работу с полной занятостью. Зарплаты здесь выше, чем при полном дне, а вакансий вполне достаточно, что б было из чего выбрать. Да и наши претенденты на позицию Data Scientist вполне могут работать удаленно (кому-то это может быть даже гараздо удобнее). Первые четыре комбинации графика и занятости я бы особо не рассматривал из-за ограниченного пула вакансий и неопределенной зарплатной вилки.

In [55]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования